lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+  /// the c-style representation
+  ///
+  /// \param StackArgumentTypes The qualified type of the relevant part of
+  ///        the stack.
+  /// \param IndexOffset The index of the first argument (indicates the register
+  ///        argument count).
+  ///
+  /// \return An ordered list of arguments.
+  static std::optional<llvm::SmallVector<model::Argument, 8>>
+  tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                              size_t IndexOffset) {
+    // Qualifiers here are not allowed since this must point to a simple struct.
     revng_assert(StackArgumentTypes.Qualifiers.empty());
+
     auto *Unqualified = StackArgumentTypes.UnqualifiedType.get();
-    if (not Unqualified)
-      return {};
+    if (!Unqualified) {
+      // If there is no type, it means that the importer responsible for this
+      // function didn't detect any stack arguments and avoided creating
+      // a new empty type.
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
 
     auto *Pointer = llvm::dyn_cast<model::StructType>(Unqualified);
     revng_assert(Pointer != nullptr,
                  "`RawFunctionType::StackArgumentsType` must be a struct");
     const model::StructType &Types = *Pointer;
 
+    // If the struct is empty, there's no arguments to extract from it.
+    if (Types.Fields.empty()) {
+      revng_assert(Types.size() == 0);
+      return llvm::SmallVector<model::Argument, 8>{};
+    }
+
+    // Compute the alignment of the first argument.
+    auto CurrentAlignment = *abi::naturalAlignment(Types.Fields.begin()->Type,
+                                                   ABI);
+    if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
+      // Abandon cases where alignment is equal to 0 or is not a power of two.
+      return std::nullopt;
+    }
+
+    // Process each of the fields, converting them into arguments separately.
     llvm::SmallVector<model::Argument, 8> Result;
-    for (const model::StructField &Field : Types.Fields) {
+    for (auto Iterator = Types.Fields.begin();
+         Iterator != std::prev(Types.Fields.end());
+         ++Iterator) {
+      const auto &CurrentArgument = *Iterator;
+      auto MaybeSize = CurrentArgument.Type.size();
+      revng_assert(MaybeSize.has_value() && MaybeSize.value() != 0);
+      uint64_t CurrentSize = paddedSizeOnStack(MaybeSize.value());
+
+      const auto &NextArgument = *std::next(Iterator);
+      auto NextAlignment = *abi::naturalAlignment(NextArgument.Type, ABI);
+      if ((CurrentAlignment & (CurrentAlignment - 1)) == 0) {
+        // Abandon cases where alignment is equal to 0 or is not a power of two.
+        return std::nullopt;
+      }
+
+      if (CurrentArgument.Offset + CurrentSize != NextArgument.Offset
+          || (CurrentArgument.Offset + CurrentSize) % NextAlignment != 0) {
+        // Abandon any cases where natural alignment causes a padding "hole"
+        // to be required.
```

Comment:
Uhm, I'm confused by this comment. The difference between `CurrentArgument.Type.size()` and `CurrentSize` takes into account this padding "hole". So we're already considering this, isn't it?

Reply:
Ah, ok, this comment is insufficient, it's my bad. It should be more explicit that this targets the "padding holes" as in "there should be an extra field in here, but there's none". As in, the padding BEFORE the potential field.

The difference you're talking about seems to be the one about the natural extension of an argument: the padding AFTER a field.

Reply:
Did this a while back but forgot to mention it here, sorry 😅.

https://github.com/revng/revng/pull/232/commits/7eaa2062e01464443c8f4f6e4a2d4aefb721651f#diff-869af126c25e96b2909972d4bfb0a8fbf2de2a1eb764a53975ddc300ee837cf5R183-R187
      // This register is unused but we still have to create an argument
      // for it. Otherwise the resulting function will not be compatible with
      // the current one: if the "hole" is not preserved, there will be no way
      // to recreate it at any point in the future, when, for example,
      // converting back to the original-like representation.
      
What do you think?

Reply:
So, in SystemV x86-64, here we're adding `rdi` if `rsi` is an argument.

> there will be no way to recreate it

This is not a goal of this analysis. The point is that it would be semantically incorrect not to put it.
Either we put it, or we say that this RFT is incompatible with this ABI, but that's probably overly strict for our goals.

Also, let's put a TODO stating that this is already done in `DetectABI` with `applyABIDeductions` and that we should consider dropping doing it there, since we're doing it again later here in any case.

Reply:
Sounds good.

Reply:
What do you think about it now? https://github.com/revng/revng/pull/232/commits/cff162fa8df2de662a1e24883267373c4b671f4c

Also, here's the TODO: https://github.com/revng/revng/pull/232/commits/635f9607e0f881e89d7abe39eb23425a5ff63d53